compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -571,64 +571,55 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
                 // Obtain the underlying trait we are working on, and the adjusted receiver argument.
                 let recv_ty = receiver.layout.ty;
-                let (vptr, dyn_ty, adjusted_receiver) = match recv_ty.kind() {
-                    ty::Ref(..) | ty::RawPtr(..)
-                        if matches!(
-                            recv_ty.builtin_deref(true).unwrap().ty.kind(),
-                            ty::Dynamic(_, _, ty::DynStar)
-                        ) =>
-                    {
-                        let receiver = self.deref_operand(&receiver)?;
-                        let ty::Dynamic(data, ..) = receiver.layout.ty.kind() else { bug!() };
-                        let (recv, vptr) = self.unpack_dyn_star(&receiver.into())?;
-                        let (dyn_ty, dyn_trait) = self.get_ptr_vtable(vptr)?;
-                        if dyn_trait != data.principal() {
-                            throw_ub_format!(
-                                "`dyn*` call on a pointer whose vtable does not match its type"
-                            );
-                        }
-                        let recv = recv.assert_mem_place(); // we passed an MPlaceTy to `unpack_dyn_star` so we definitely still have one
-
-                        (vptr, dyn_ty, recv.ptr)
-                    }
+                let receiver_place = match recv_ty.kind() {
+                    ty::Ref(..) | ty::RawPtr(..) => self.deref_operand(&receiver)?,
+                    ty::Dynamic(_, _, ty::Dyn) => receiver.assert_mem_place(), // unsized (`dyn`) cannot be immediate
```

Comment:
I thought I had a reason for doing things in this particular way, but I can't see any case where your patch changes behavior so probably I was just blind...

Doesn't this mean we can go back even closer to how things were before this PR, and do the `deref_operand`/`assert_mem_place` in the loop above?

Reply:
yea, that works

Reply:
Nice, thanks. :)